// continuo/lib/js/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _jquery = require('jquery');

var _jquery2 = _interopRequireDefault(_jquery);

var _backbone = require('backbone');

var Backbone = _interopRequireWildcard(_backbone);

var _verovioInteractionView = require('./views/verovioInteractionView');

var _verovioInteractionView2 = _interopRequireDefault(_verovioInteractionView);

var _highlightView = require('./views/highlightView');

var _highlightView2 = _interopRequireDefault(_highlightView);

var _modelMEIdata = require('./data/model-MEIdata');

var _modelMEIdata2 = _interopRequireDefault(_modelMEIdata);

var _backboneEvents = require('./utils/backbone-events');

var _backboneEvents2 = _interopRequireDefault(_backboneEvents);

var _verovioExt = require('./utils/verovio-ext');

var _verovioExt2 = _interopRequireDefault(_verovioExt);

var _EMAexpr = require('./components/EMAexpr');

var _EMAexpr2 = _interopRequireDefault(_EMAexpr);

require('xmldom');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// NOTES
// jQuery is only used for HTML DOM operations. All XML DOM operations require xmldom.

var Continuo = function (_Backbone$View) {
    _inherits(Continuo, _Backbone$View);

    function Continuo() {
        _classCallCheck(this, Continuo);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(Continuo).apply(this, arguments));
    }

    _createClass(Continuo, [{
        key: 'initialize',
        value: function initialize(options) {
            this.mei = options.mei;
            this.meiString = options.meiString;
            this.omas = options.omas;
            this.listenTo(_backboneEvents2.default, 'component:emaBox', this.updateEmaBox);
            this.listenTo(_backboneEvents2.default, 'addFile', this.addFile);
        }
    }, {
        key: 'updateEmaBox',
        value: function updateEmaBox(expr) {
            (0, _jquery2.default)(".cnt-emabox").show();
            (0, _jquery2.default)(".cnt-emabox").text(expr);
        }
    }, {
        key: 'addFile',
        value: function addFile(textData) {
            // Create score
            var container = this.$el.find(".cnt-container");
            _backboneEvents2.default.trigger('component:emaBox:url', textData["url"]);

            var doc = new DOMParser().parseFromString(textData["string"], 'text/xml');
            this.MEIdata = new _modelMEIdata2.default({ "doc": doc,
                "string": textData["string"]
            });
            this.MEIdata.generate_ids();

            // Sadly, importing Verovio crashes babelify,
            // so we assume it's globally available
            // i.e. verovio must be defined.
            var vrvToolkit = new verovio.toolkit();
            var scale = 50;
            var options = JSON.stringify({
                pageWidth: this.$el.width() * 100 / scale,
                ignoreLayout: 1,
                adjustPageHeight: 1,
                border: 50,
                scale: scale
            });
            vrvToolkit.setOptions(options);
            var mei = textData["string"];
            vrvToolkit.loadData(mei + "\n", "");
            var pgs = vrvToolkit.getPageCount();
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = Array.from(new Array(pgs), function (x, i) {
                    return i;
                })[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var page = _step.value;

                    var svg = vrvToolkit.renderPage(page + 1);

                    var ext_svg = (0, _verovioExt2.default)(svg);
                    container.append(ext_svg);
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            new _verovioInteractionView2.default({ "el": container, "model": this.MEIdata });

            // Determine if notation must be highlighted
            var api_opts = textData["url"].split("/").pop();
            if (api_opts.includes("highlight")) {
                new _highlightView2.default({ "el": container, "model": this.MEIdata });
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _this2 = this;

            var container = (0, _jquery2.default)("<div class='cnt-container'></div>");
            this.$el.append(container);

            if (this.meiString) {
                this.addFile({
                    "filename": "",
                    "url": "",
                    "string": this.meiString
                });
            } else if (this.mei) {
                var url = this.mei;

                if (this.omas) {
                    url = this.omas + encodeURIComponent(this.mei) + "/all/all/@all";
                }

                _jquery2.default.ajax({
                    url: url,
                    type: 'GET',
                    dataType: 'text',
                    success: function success(data) {
                        _this2.addFile({
                            "filename": _this2.mei,
                            "url": _this2.mei,
                            "string": data
                        });
                    }
                });
            }
            // else {
            //     // Create controls floating box
            //     let controls = $("<div class='cnt-controls'></div>");
            //     container.append(controls);
            //
            //     // Render components (model-less subviews)
            //     // new FileUploadComponent({"el":controls});
            //     new DropboxUploadComponent({"el":controls});
            //     new FileFromWebComponent({"el":controls});
            // }

            // Create EMA floating box
            container.append(new _EMAexpr2.default().render());
        }
    }]);

    return Continuo;
}(Backbone.View);

// Make main class available to pre-ES6 browser environments


if (window) {
    window.Continuo = Continuo;
}
exports.default = Continuo;
