// continuo/lib/js/components/EMAexpr.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _jquery = require('jquery');

var _jquery2 = _interopRequireDefault(_jquery);

var _backbone = require('backbone');

var Backbone = _interopRequireWildcard(_backbone);

var _backboneEvents = require('../utils/backbone-events');

var _backboneEvents2 = _interopRequireDefault(_backboneEvents);

var _emaexprTpl = require('../templates/emaexpr-tpl');

var _emaexprTpl2 = _interopRequireDefault(_emaexprTpl);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EMAExprComponent = function (_Backbone$View) {
    _inherits(EMAExprComponent, _Backbone$View);

    function EMAExprComponent() {
        _classCallCheck(this, EMAExprComponent);

        return _possibleConstructorReturn(this, (EMAExprComponent.__proto__ || Object.getPrototypeOf(EMAExprComponent)).apply(this, arguments));
    }

    _createClass(EMAExprComponent, [{
        key: 'template',
        value: function template(tpl) {
            return (0, _emaexprTpl2.default)(tpl);
        }
    }, {
        key: 'initialize',
        value: function initialize() {
            var _this2 = this;

            this.on('component:emaBox', this.updateEmaBox);
            this.on("component:emaBox:url", function (u) {
                _this2.MEIurl = u;
            });
        }
    }, {
        key: 'copyExpr',
        value: function copyExpr(e) {
            e.preventDefault();
            var url = "/" + encodeURIComponent(this.MEIurl) + "/" + this.expr;
            window.prompt("Copy to clipboard: Ctrl+C, Enter", url);
        }
    }, {
        key: 'toggle',
        value: function toggle(e) {
            e.preventDefault();
            if (this.$el.hasClass("cnt-hidden")) {
                this.$el.removeClass("cnt-hidden");
            } else {
                this.$el.addClass("cnt-hidden");
            }
        }
    }, {
        key: 'updateEmaBox',
        value: function updateEmaBox(expr) {
            this.$el.show();
            this.expr = expr;
            // let server = "http://localhost:5000";
            var server = "http://159.65.177.99:5000";
            var loc = window.location.origin + window.location.pathname;
            var tpl_data = {
                "expr": expr,
                "url": loc + "#" + server + "/" + encodeURIComponent(this.MEIurl) + "/" + this.expr + "/highlight"
            };
            this.$el.html(this.template(tpl_data));
            this.delegateEvents(this.events);
            if (expr == "") {
                this.$el.hide();
            }
        }
    }, {
        key: 'render',
        value: function render() {
            return this.$el;
        }
    }, {
        key: 'className',
        get: function get() {
            return "cnt-emabox";
        }
    }, {
        key: 'events',
        get: function get() {
            var _this3 = this;

            return {
                "click .cnt-emabox-x": function clickCntEmaboxX() {
                    _this3.trigger("component:emaBox:clear");
                },
                "click .cnt-emabox-cp": this.copyExpr,
                "click .cnt-emabox-hide": this.toggle
            };
        }
    }]);

    return EMAExprComponent;
}(Backbone.View);

exports.default = EMAExprComponent;
